File: frontend/src/types/history.ts
/** History, analytics, and A/B test types (Phase 6). */

// ── Execution ───────────────────────────────────────────

export interface ExecutionSummary {
  id: string;
  session_id: string;
  model: string;
  input_text: string;
  result_text: string | null;
  success: boolean;
  status: 'running' | 'completed' | 'error' | 'cancelled';
  iterations: number;
  total_cost_usd: number;
  duration_ms: number;
  total_tokens: number;
  tool_call_count: number;
  environment_id: string | null;
  created_at: string;
  tags: string[];
}

export interface StageTiming {
  iteration: number;
  stage_order: number;
  stage_name: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  input_tokens: number;
  output_tokens: number;
  was_cached: boolean;
  was_skipped: boolean;
  tool_name: string | null;
  tool_success: boolean | null;
  tool_duration_ms: number | null;
}

export interface ToolCallRecord {
  iteration: number;
  tool_name: string;
  called_at: string;
  input_json: string | null;
  output_text: string | null;
  is_error: boolean;
  duration_ms: number;
}

export interface ExecutionDetail extends ExecutionSummary {
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
  thinking_tokens: number;
  error_type: string | null;
  error_message: string | null;
  error_stage: number | null;
  finished_at: string | null;
  stage_timings: StageTiming[];
  tool_call_records: ToolCallRecord[];
}

// ── Waterfall ───────────────────────────────────────────

export interface StageWaterfall {
  order: number;
  name: string;
  duration_ms: number;
  was_cached: boolean;
  was_skipped: boolean;
  tokens: number;
}

export interface IterationWaterfall {
  iteration: number;
  stages: StageWaterfall[];
}

export interface WaterfallData {
  execution_id: string;
  total_duration_ms: number;
  iterations: IterationWaterfall[];
}

// ── Stage Stats ─────────────────────────────────────────

export interface StageStats {
  stage_order: number;
  stage_name: string;
  count: number;
  avg_ms: number;
  min_ms: number;
  max_ms: number;
  cache_hits: number;
  skips: number;
  avg_input_tokens: number;
  avg_output_tokens: number;
}

// ── Cost Analysis ───────────────────────────────────────

export interface ModelCost {
  model: string;
  executions: number;
  total_cost: number;
  total_tokens: number;
  total_input: number;
  total_output: number;
  total_cache_read: number;
  total_cache_write: number;
  total_thinking: number;
  total_tools: number;
  avg_cost: number;
}

export interface CostSummary {
  session_id: string | null;
  by_model: ModelCost[];
  total_cost: number;
  total_executions: number;
}

export interface CostTrendPoint {
  period: string;
  executions: number;
  cost: number;
  tokens: number;
}

// ── A/B Test ────────────────────────────────────────────

export interface ABSide {
  execution_id: string;
  model: string;
  status: string;
  result_text: string;
  cost_usd: number;
  duration_ms: number;
  total_tokens: number;
  iterations: number;
  tool_calls: number;
}

export interface ABComparison {
  env_a: ABSide;
  env_b: ABSide;
  diff: {
    cost_diff: number;
    duration_diff: number;
    token_diff: number;
  };
}

// ── Global Stats ────────────────────────────────────────

export interface HistoryStats {
  total: number;
  completed: number;
  errors: number;
  total_cost: number;
  total_tokens: number;
  avg_duration_ms: number;
}

File: backend/app/schemas/history.py
"""History, analytics, and A/B test API schemas (Phase 6)."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel


# ── Execution list / detail ──────────────────────────────


class ExecutionSummary(BaseModel):
    id: str
    session_id: str
    model: str = ""
    input_text: str = ""
    result_text: Optional[str] = None
    success: bool = True
    status: str = "completed"
    iterations: int = 0
    total_cost_usd: float = 0.0
    duration_ms: int = 0
    total_tokens: int = 0
    tool_call_count: int = 0
    environment_id: Optional[str] = None
    created_at: str = ""
    tags: List[str] = []


class ExecutionListResponse(BaseModel):
    runs: List[ExecutionSummary]
    total: int = 0


class StageTimingResponse(BaseModel):
    iteration: int
    stage_order: int
    stage_name: str
    started_at: str
    finished_at: str
    duration_ms: int
    input_tokens: int = 0
    output_tokens: int = 0
    was_cached: bool = False
    was_skipped: bool = False
    tool_name: Optional[str] = None
    tool_success: Optional[bool] = None
    tool_duration_ms: Optional[int] = None


class ToolCallResponse(BaseModel):
    iteration: int
    tool_name: str
    called_at: str
    input_json: Optional[str] = None
    output_text: Optional[str] = None
    is_error: bool = False
    duration_ms: int = 0


class ExecutionDetailResponse(BaseModel):
    id: str
    session_id: str
    model: str = ""
    input_text: str = ""
    result_text: Optional[str] = None
    success: bool = True
    status: str = "completed"
    iterations: int = 0
    total_cost_usd: float = 0.0
    duration_ms: int = 0
    total_tokens: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    thinking_tokens: int = 0
    tool_call_count: int = 0
    environment_id: Optional[str] = None
    error_type: Optional[str] = None
    error_message: Optional[str] = None
    error_stage: Optional[int] = None
    created_at: str = ""
    finished_at: Optional[str] = None
    tags: List[str] = []
    stage_timings: List[StageTimingResponse] = []
    tool_call_records: List[ToolCallResponse] = []


# ── Waterfall ────────────────────────────────────────────


class StageWaterfallResponse(BaseModel):
    order: int
    name: str
    duration_ms: int
    was_cached: bool = False
    was_skipped: bool = False
    tokens: int = 0


class IterationWaterfallResponse(BaseModel):
    iteration: int
    stages: List[StageWaterfallResponse] = []


class WaterfallResponse(BaseModel):
    execution_id: str
    total_duration_ms: int = 0
    iterations: List[IterationWaterfallResponse] = []


# ── Stage Stats ──────────────────────────────────────────


class StageStatsResponse(BaseModel):
    stage_order: int
    stage_name: str
    count: int
    avg_ms: float
    min_ms: float
    max_ms: float
    cache_hits: int = 0
    skips: int = 0
    avg_input_tokens: float = 0.0
    avg_output_tokens: float = 0.0


# ── Cost Analysis ────────────────────────────────────────


class ModelCostResponse(BaseModel):
    model: str
    executions: int = 0
    total_cost: float = 0.0
    total_tokens: int = 0
    total_input: int = 0
    total_output: int = 0
    total_cache_read: int = 0
    total_cache_write: int = 0
    total_thinking: int = 0
    total_tools: int = 0
    avg_cost: float = 0.0


class CostSummaryResponse(BaseModel):
    session_id: Optional[str] = None
    by_model: List[ModelCostResponse] = []
    total_cost: float = 0.0
    total_executions: int = 0


class CostTrendPointResponse(BaseModel):
    period: str
    executions: int = 0
    cost: float = 0.0
    tokens: int = 0


class CostTrendResponse(BaseModel):
    trend: List[CostTrendPointResponse] = []
    granularity: str = "hour"


# ── Stats ────────────────────────────────────────────────


class StatsResponse(BaseModel):
    total: int = 0
    completed: int = 0
    errors: int = 0
    total_cost: float = 0.0
    total_tokens: int = 0
    avg_duration_ms: float = 0.0


# ── A/B Test ─────────────────────────────────────────────


class ABTestRequest(BaseModel):
    env_a_id: str
    env_b_id: str
    user_input: str


class ABSideResponse(BaseModel):
    execution_id: str
    model: str = ""
    status: str = ""
    result_text: str = ""
    cost_usd: float = 0.0
    duration_ms: int = 0
    total_tokens: int = 0
    iterations: int = 0
    tool_calls: int = 0


class ABComparisonResponse(BaseModel):
    env_a: ABSideResponse
    env_b: ABSideResponse
    diff: Dict[str, Any] = {}
